diff --git a/drivers/cpufreq/Kconfig b/drivers/cpufreq/Kconfig
index d49ac2073..bd5692aa5 100644
--- a/drivers/cpufreq/Kconfig
+++ b/drivers/cpufreq/Kconfig
@@ -203,6 +203,65 @@ config CPU_BOOST
 
 	  If in doubt, say N.
 
+config CPU_INPUT_BOOST
+	bool "CPU Input Boost"
+	help
+	  Boosts the CPU on touchscreen and touchpad input, and allows for
+	  boosting on other custom events, mainly which is intended to be for
+	  boosting when there is a new frame ready to be rendered to the
+	  display. The boost frequencies for this driver should be set so that
+	  frame drops are near-zero at the boosted frequencies and power
+	  consumption is minimized at said frequency combination.
+
+if CPU_INPUT_BOOST
+
+config MAX_BOOST_FREQ_LP
+	int "Low-power cluster max-boost freq"
+	default "0"
+	help
+	  Max-boost frequency for the low-power CPU cluster.
+
+config MAX_BOOST_FREQ_PERF
+	int "Performance cluster max-boost freq"
+	default "0"
+	help
+	  Max-boost frequency for the performance CPU cluster.
+
+config MAX_BOOST_FREQ_PRIME
+	int "Prime cluster max-boost freq"
+	default "0"
+	help
+	  Max-boost frequency for the prime CPU cluster.
+
+config CPU_FREQ_MIN_LP
+	int "Default minimum frequency for the little cluster"
+	default 0
+	help
+	  This sets the default minimum frequency (in kHz) for the little CPU
+	  cluster.
+
+	  If in doubt, say 0 to use the hardware's minimum frequency.
+
+config CPU_FREQ_MIN_PERF
+	int "Default minimum frequency for the big cluster"
+	default 0
+	help
+	  This sets the default minimum frequency (in kHz) for the big CPU
+	  cluster.
+
+	  If in doubt, say 0 to use the hardware's minimum frequency.
+
+config CPU_FREQ_MIN_PRIME
+	int "Default minimum frequency for the prime cluster"
+	default 0
+	help
+	  This sets the default minimum frequency (in kHz) for the prime CPU
+	  cluster.
+
+	  If in doubt, say 0 to use the hardware's minimum frequency.
+
+endif
+
 config CPU_FREQ_GOV_SCHEDUTIL
 	bool "'schedutil' cpufreq policy governor"
 	depends on CPU_FREQ && SMP
diff --git a/drivers/cpufreq/Makefile b/drivers/cpufreq/Makefile
index 9dfd353f0..cc299fb3b 100644
--- a/drivers/cpufreq/Makefile
+++ b/drivers/cpufreq/Makefile
@@ -18,6 +18,9 @@ obj-$(CONFIG_CPU_FREQ_GOV_COMMON)		+= cpufreq_governor.o
 obj-$(CONFIG_CPU_FREQ_GOV_ATTR_SET)	+= cpufreq_governor_attr_set.o
 obj-$(CONFIG_CPU_BOOST)			+= cpu-boost.o
 
+# CPU Input Boost
+obj-$(CONFIG_CPU_INPUT_BOOST)		+= cpu_input_boost.o
+
 obj-$(CONFIG_CPUFREQ_DT)		+= cpufreq-dt.o
 obj-$(CONFIG_CPUFREQ_DT_PLATDEV)	+= cpufreq-dt-platdev.o
 
diff --git a/drivers/cpufreq/cpu-boost.c b/drivers/cpufreq/cpu-boost.c
index cb07beebd..49d8c5b61 100644
--- a/drivers/cpufreq/cpu-boost.c
+++ b/drivers/cpufreq/cpu-boost.c
@@ -9,11 +9,15 @@
 #include <linux/init.h>
 #include <linux/cpufreq.h>
 #include <linux/cpu.h>
+#include <linux/kthread.h>
 #include <linux/sched.h>
 #include <linux/slab.h>
 #include <linux/input.h>
 #include <linux/time.h>
 #include <linux/sysfs.h>
+#include <uapi/linux/sched/types.h>
+
+#include <linux/sched/rt.h>
 
 #define cpu_boost_attr_rw(_name)		\
 static struct kobj_attribute _name##_attr =	\
@@ -43,9 +47,8 @@ struct cpu_sync {
 };
 
 static DEFINE_PER_CPU(struct cpu_sync, sync_info);
-static struct workqueue_struct *cpu_boost_wq;
 
-static struct work_struct input_boost_work;
+static struct kthread_work input_boost_work;
 
 static bool input_boost_enabled;
 
@@ -63,7 +66,11 @@ static bool sched_boost_active;
 
 static struct delayed_work input_boost_rem;
 static u64 last_input_time;
-#define MIN_INPUT_INTERVAL (150 * USEC_PER_MSEC)
+
+static struct kthread_worker cpu_boost_worker;
+static struct task_struct *cpu_boost_worker_thread;
+
+#define MIN_INPUT_INTERVAL (100 * USEC_PER_MSEC)
 
 static ssize_t store_input_boost_freq(struct kobject *kobj,
 				      struct kobj_attribute *attr,
@@ -203,7 +210,7 @@ static void do_input_boost_rem(struct work_struct *work)
 	}
 }
 
-static void do_input_boost(struct work_struct *work)
+static void do_input_boost(struct kthread_work *work)
 {
 	unsigned int i, ret;
 	struct cpu_sync *i_sync_info;
@@ -233,8 +240,7 @@ static void do_input_boost(struct work_struct *work)
 			sched_boost_active = true;
 	}
 
-	queue_delayed_work(cpu_boost_wq, &input_boost_rem,
-					msecs_to_jiffies(input_boost_ms));
+	schedule_delayed_work(&input_boost_rem, msecs_to_jiffies(input_boost_ms));
 }
 
 static void cpuboost_input_event(struct input_handle *handle,
@@ -249,10 +255,10 @@ static void cpuboost_input_event(struct input_handle *handle,
 	if (now - last_input_time < MIN_INPUT_INTERVAL)
 		return;
 
-	if (work_pending(&input_boost_work))
+	if (queuing_blocked(&cpu_boost_worker, &input_boost_work))
 		return;
 
-	queue_work(cpu_boost_wq, &input_boost_work);
+	kthread_queue_work(&cpu_boost_worker, &input_boost_work);
 	last_input_time = ktime_to_us(ktime_get());
 }
 
@@ -330,14 +336,36 @@ static struct input_handler cpuboost_input_handler = {
 struct kobject *cpu_boost_kobj;
 static int cpu_boost_init(void)
 {
-	int cpu, ret;
+	int cpu, ret, i;
 	struct cpu_sync *s;
+	struct sched_param param = { .sched_priority = 2 };
+	cpumask_t sys_bg_mask;
+
+	/* Hardcode the cpumask to bind the kthread to it */
+	cpumask_clear(&sys_bg_mask);
+	for (i = 0; i <= 3; i++) {
+		cpumask_set_cpu(i, &sys_bg_mask);
+	}
 
-	cpu_boost_wq = alloc_workqueue("cpuboost_wq", WQ_HIGHPRI, 0);
-	if (!cpu_boost_wq)
+	kthread_init_worker(&cpu_boost_worker);
+	cpu_boost_worker_thread = kthread_create(kthread_worker_fn,
+		&cpu_boost_worker, "cpu_boost_worker_thread");
+	if (IS_ERR(cpu_boost_worker_thread)) {
+		pr_err("cpu-boost: Failed to init kworker!\n");
 		return -EFAULT;
+	}
+
+	ret = sched_setscheduler(cpu_boost_worker_thread, SCHED_FIFO, &param);
+	if (ret)
+		pr_err("cpu-boost: Failed to set SCHED_FIFO!\n");
+
+	/* Now bind it to the cpumask */
+	kthread_bind_mask(cpu_boost_worker_thread, &sys_bg_mask);
+
+	/* Wake it up! */
+	wake_up_process(cpu_boost_worker_thread);
 
-	INIT_WORK(&input_boost_work, do_input_boost);
+	kthread_init_work(&input_boost_work, do_input_boost);
 	INIT_DELAYED_WORK(&input_boost_rem, do_input_boost_rem);
 
 	for_each_possible_cpu(cpu) {
diff --git a/drivers/cpufreq/cpu_input_boost.c b/drivers/cpufreq/cpu_input_boost.c
new file mode 100644
index 000000000..c217b8f10
--- /dev/null
+++ b/drivers/cpufreq/cpu_input_boost.c
@@ -0,0 +1,259 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2018-2019 Sultan Alsawaf <sultan@kerneltoast.com>.
+ */
+
+#define pr_fmt(fmt) "cpu_input_boost: " fmt
+
+#include <linux/cpu.h>
+#include <linux/cpufreq.h>
+#include <linux/input.h>
+#include <linux/kthread.h>
+#include <linux/moduleparam.h>
+#include <drm/drm_notifier_mi.h>
+#include <linux/slab.h>
+#include <linux/version.h>
+
+/* The sched_param struct is located elsewhere in newer kernels */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 10, 0)
+#include <uapi/linux/sched/types.h>
+#endif
+
+static unsigned int max_boost_freq_little __read_mostly =
+	CONFIG_MAX_BOOST_FREQ_LP;
+static unsigned int max_boost_freq_big __read_mostly =
+	CONFIG_MAX_BOOST_FREQ_PERF;
+static unsigned int max_boost_freq_prime __read_mostly =
+	CONFIG_MAX_BOOST_FREQ_PRIME;
+static unsigned int cpu_freq_min_little __read_mostly =
+	CONFIG_CPU_FREQ_MIN_LP;
+static unsigned int cpu_freq_min_big __read_mostly =
+	CONFIG_CPU_FREQ_MIN_PERF;
+static unsigned int cpu_freq_min_prime __read_mostly =
+	CONFIG_CPU_FREQ_MIN_PRIME;
+
+module_param(max_boost_freq_little, uint, 0644);
+module_param(max_boost_freq_big, uint, 0644);
+module_param(max_boost_freq_prime, uint, 0644);
+module_param(cpu_freq_min_little, uint, 0644);
+module_param(cpu_freq_min_big, uint, 0644);
+module_param(cpu_freq_min_prime, uint, 0644);
+
+enum {
+	SCREEN_OFF,
+	MAX_BOOST
+};
+
+struct boost_drv {
+	struct delayed_work max_unboost;
+	struct notifier_block cpu_notif;
+	struct notifier_block msm_drm_notif;
+	wait_queue_head_t boost_waitq;
+	atomic_long_t max_boost_expires;
+	unsigned long state;
+};
+
+static void max_unboost_worker(struct work_struct *work);
+
+static struct boost_drv boost_drv_g __read_mostly = {
+	.max_unboost = __DELAYED_WORK_INITIALIZER(boost_drv_g.max_unboost,
+						  max_unboost_worker, 0),
+	.boost_waitq = __WAIT_QUEUE_HEAD_INITIALIZER(boost_drv_g.boost_waitq)
+};
+
+static unsigned int get_max_boost_freq(struct cpufreq_policy *policy)
+{
+	unsigned int freq;
+
+	if (cpumask_test_cpu(policy->cpu, cpu_lp_mask))
+		freq = max_boost_freq_little;
+	else if (cpumask_test_cpu(policy->cpu, cpu_perf_mask))
+		freq = max_boost_freq_big;
+	else
+		freq = max_boost_freq_prime;
+	return min(freq, policy->max);
+}
+
+static unsigned int get_min_freq(struct cpufreq_policy *policy)
+{
+	unsigned int freq;
+
+	if (cpumask_test_cpu(policy->cpu, cpu_lp_mask))
+		freq = cpu_freq_min_little;
+	else if (cpumask_test_cpu(policy->cpu, cpu_perf_mask))
+		freq = cpu_freq_min_big;
+	else
+		freq = cpu_freq_min_prime;
+
+	return max(freq, policy->cpuinfo.min_freq);
+}
+
+
+static void update_online_cpu_policy(void)
+{
+	unsigned int cpu;
+
+	/* Only one CPU from each cluster needs to be updated */
+	get_online_cpus();
+	cpu = cpumask_first_and(cpu_lp_mask, cpu_online_mask);
+	cpufreq_update_policy(cpu);
+	cpu = cpumask_first_and(cpu_perf_mask, cpu_online_mask);
+	cpufreq_update_policy(cpu);
+	cpu = cpumask_first_and(cpu_prime_mask, cpu_online_mask);
+	cpufreq_update_policy(cpu);
+	put_online_cpus();
+}
+
+static void __cpu_input_boost_kick_max(struct boost_drv *b,
+				       unsigned int duration_ms)
+{
+	unsigned long boost_jiffies = msecs_to_jiffies(duration_ms);
+	unsigned long curr_expires, new_expires;
+
+	if (test_bit(SCREEN_OFF, &b->state))
+		return;
+
+	do {
+		curr_expires = atomic_long_read(&b->max_boost_expires);
+		new_expires = jiffies + boost_jiffies;
+
+		/* Skip this boost if there's a longer boost in effect */
+		if (time_after(curr_expires, new_expires))
+			return;
+	} while (atomic_long_cmpxchg(&b->max_boost_expires, curr_expires,
+				     new_expires) != curr_expires);
+
+	set_bit(MAX_BOOST, &b->state);
+	if (!mod_delayed_work(system_unbound_wq, &b->max_unboost,
+			      boost_jiffies))
+		wake_up(&b->boost_waitq);
+}
+
+void cpu_input_boost_kick_max(unsigned int duration_ms)
+{
+	struct boost_drv *b = &boost_drv_g;
+
+	__cpu_input_boost_kick_max(b, duration_ms);
+}
+
+static void max_unboost_worker(struct work_struct *work)
+{
+	struct boost_drv *b = container_of(to_delayed_work(work),
+					   typeof(*b), max_unboost);
+
+	clear_bit(MAX_BOOST, &b->state);
+	wake_up(&b->boost_waitq);
+}
+
+static int cpu_boost_thread(void *data)
+{
+	static const struct sched_param sched_max_rt_prio = {
+		.sched_priority = MAX_RT_PRIO - 1
+	};
+	struct boost_drv *b = data;
+	unsigned long old_state = 0;
+
+	sched_setscheduler_nocheck(current, SCHED_FIFO, &sched_max_rt_prio);
+
+	while (1) {
+		bool should_stop = false;
+		unsigned long curr_state;
+
+		wait_event(b->boost_waitq,
+			(curr_state = READ_ONCE(b->state)) != old_state ||
+			(should_stop = kthread_should_stop()));
+
+		if (should_stop)
+			break;
+
+		old_state = curr_state;
+		update_online_cpu_policy();
+	}
+
+	return 0;
+}
+
+static int cpu_notifier_cb(struct notifier_block *nb, unsigned long action,
+			   void *data)
+{
+	struct boost_drv *b = container_of(nb, typeof(*b), cpu_notif);
+	struct cpufreq_policy *policy = data;
+
+	if (action != CPUFREQ_ADJUST)
+		return NOTIFY_OK;
+
+	/* Unboost when the screen is off */
+	if (test_bit(SCREEN_OFF, &b->state)) {
+		policy->min = get_min_freq(policy);
+		return NOTIFY_OK;
+	}
+
+	/* Boost CPU to max frequency for max boost */
+	if (test_bit(MAX_BOOST, &b->state)) {
+		policy->min = get_max_boost_freq(policy);
+		return NOTIFY_OK;
+	}
+
+	policy->min = get_min_freq(policy);
+
+	return NOTIFY_OK;
+}
+
+static int msm_drm_notifier_cb(struct notifier_block *nb, unsigned long action,
+			  void *data)
+{
+	struct boost_drv *b = container_of(nb, typeof(*b), msm_drm_notif);
+	int *blank = ((struct mi_drm_notifier *)data)->data;
+
+	/* Parse framebuffer blank events as soon as they occur */
+	if (action != MI_DRM_EARLY_EVENT_BLANK)
+		return NOTIFY_OK;
+
+	/* Unboost when screen turns off */
+	if (*blank == MI_DRM_BLANK_UNBLANK) {
+		clear_bit(SCREEN_OFF, &b->state);
+	} else {
+		set_bit(SCREEN_OFF, &b->state);
+		wake_up(&b->boost_waitq);
+	}
+
+	return NOTIFY_OK;
+}
+
+static int __init cpu_input_boost_init(void)
+{
+	struct boost_drv *b = &boost_drv_g;
+	struct task_struct *thread;
+	int ret;
+
+	b->cpu_notif.notifier_call = cpu_notifier_cb;
+	ret = cpufreq_register_notifier(&b->cpu_notif, CPUFREQ_POLICY_NOTIFIER);
+	if (ret) {
+		pr_err("Failed to register cpufreq notifier, err: %d\n", ret);
+		return ret;
+	}
+
+	b->msm_drm_notif.notifier_call = msm_drm_notifier_cb;
+	b->msm_drm_notif.priority = INT_MAX;
+	ret = mi_drm_register_client(&b->msm_drm_notif);
+	if (ret) {
+		pr_err("Unable to register mi_drm notifier: %d\n", ret);
+		goto unregister_cpu_notif;
+	}
+
+	thread = kthread_run(cpu_boost_thread, b, "cpu_boostd");
+	if (IS_ERR(thread)) {
+		ret = PTR_ERR(thread);
+		pr_err("Failed to start CPU boost thread, err: %d\n", ret);
+		goto unregister_fb_notif;
+	}
+
+	return 0;
+
+unregister_fb_notif:
+	mi_drm_unregister_client(&b->msm_drm_notif);
+unregister_cpu_notif:
+	cpufreq_unregister_notifier(&b->cpu_notif, CPUFREQ_POLICY_NOTIFIER);
+	return ret;
+}
+late_initcall(cpu_input_boost_init);
